pytorch_lightning/plugins/collective/collective_plugin.py

```diff
@@ -0,0 +1,74 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional, Union
+
+import torch
+
+from pytorch_lightning.utilities.distributed import ReduceOp
+
+
+class CollectivePlugin(ABC):
+    """Interface for collective functions for training type plugins.
```

Comment:
i don't think we ought to mention training type plugins here. it should be as self-contained as possible, though within Lightning these will be managed by the training type plugins.

A good test would be that someone could use these utilities independently of the rest of lightning if they wanted to. that means it'll be easier for us to integrate